satellite/overlay/combinedcache.go

```diff
@@ -0,0 +1,141 @@
+// Copyright (C) 2019 Storj Labs, Inc.
+// See LICENSE for copying information.
+
+package overlay
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"storj.io/storj/pkg/pb"
+	"storj.io/storj/pkg/storj"
+)
+
+type addressInfo struct {
+	address   string
+	lastIP    string
+	transport pb.NodeTransport
+}
+
+type uptimeInfo struct {
+	isUp       bool
+	lastUptime time.Time
+	stats      *NodeStats
+}
+
+// CombinedCache is a simple caching mechanism for overlaycache updates. It
+// provdes methods to help reduce calls to UpdateAddress and UpdateTime, but can
+// be extended for other calls in the future.
+type CombinedCache struct {
+	DB
+	addressLock  sync.RWMutex
+	addressCache map[storj.NodeID]*addressInfo
+
+	uptimeLock          sync.RWMutex
+	uptimeCache         map[storj.NodeID]*uptimeInfo
+	uptimeFlushInterval time.Duration
+}
+
+// NewCombinedCache instantiates a new CombinedCache
+func NewCombinedCache(db DB, uptimeFlushInterval time.Duration) *CombinedCache {
+	return &CombinedCache{
+		DB:                  db,
+		addressCache:        make(map[storj.NodeID]*addressInfo),
+		uptimeCache:         make(map[storj.NodeID]*uptimeInfo),
+		uptimeFlushInterval: uptimeFlushInterval,
+	}
+}
+
+// UpdateUptime overrides the underlying db.UpdateUptime and provides a simple
+// caching layer to reduce calls to the underlying db.
+func (c *CombinedCache) UpdateUptime(ctx context.Context, nodeID storj.NodeID, isUp bool, lambda, weight, uptimeDQ float64) (stats *NodeStats, err error) {
+	// First check the internal cache. If it returns stats, use them.
+	stats = c.GetNodeStats(nodeID, isUp)
+	if stats != nil {
+		return stats, nil
```

Comment:
I'm still worrying about potential harm from throwing away the new information here. With the underlying overlayCache as it is, we definitely want to pass on the uptime to `c.DB.UpdateUptime()` so it can adjust the node stats, which includes a whole bunch of things other than last contact success and last contact failure. I know team green is changing all of how that works, but I expect we'll still want as much data about when a node was up and when it was down as we can get.

It is a valid option just to merge what we have here and then change it later if needed, since we don't really know exactly what team green plans, but if we do that I recommend adding to the docstring here to make it very explicitly clear that we will throw away the values of `lambda`, `weight`, and `uptimeDQ` if there was already an update recently.